### DISCUSSION TITLE
Is the (un)biased AD handled already?

### DISCUSSION THREAD

**Discussion by saeedhd96** [mitsuba-renderer/mitsuba3#371] [Q&A]
Hi,

I recall in Mitsuba2, for optimizing using pure AD, we needed to do the primal mode twice, and replace the AD graph in the second pirmal mode before doing backward mode.

Is this already handled in this version?

The use case is in integrator 'path' and calling dr.backward(loss).

Thanks

**Comment by ziyi-zhang** [comment]
Hi Saeed, 
This is handled by `mi.render()` customOP defined [here](https://github.com/mitsuba-renderer/mitsuba3/blob/2d7734731f902002adcb272fb7c08688e8ef64db/src/python/python/util.py#L405). An example of `render_for/backward()` can be found [here](https://github.com/mitsuba-renderer/mitsuba3/blob/2d7734731f902002adcb272fb7c08688e8ef64db/src/python/python/ad/integrators/common.py#L8).
But we can still run into that issue if we manually call `render_backward` without specifying a different seed.
